### ISSUE TITLE
flatten is not inferring type properly

### ISSUE THREAD

**Issue by AlexGalays** [staltz/xstream#4]
Hello,

Using Typescript 1.8.7,

I have `stream.map(fnThatReturnsAStreamOfBooleans)`. That properly get inferred to `Stream<Stream<boolean>>`

However, if I finish the chain with flatten(), the final type is `Stream<{}>` instead of `Stream<boolean>`

Example:

```
const stream: Stream<string> = Stream.create({
	start: listener => {},
	stop: () => {}
})

const stream2: Stream<boolean> = Stream.create({
	start: listener => {},
	stop: () => {}
})

const f = stream.map(x => stream2).flatten()

f is a stream<{}>
```

I suspect the `T` in flatten completely redefines the `T` of the Stream.

I think it would be best to remove this whole `T extends Stream<R>` for now.

PS: this seems to be related:  https://github.com/Microsoft/TypeScript/issues/1290

Arguably, this shows a single operation flatMap is not such a bad idea after all :p

**Comment by staltz** [comment]
Thanks for the report

**Comment by staltz** [comment]
Fixed in 2.1.0
